src/Core/Framework/DataAbstractionLayer/Doctrine/RetryableQuery.php

```diff
@@ -17,38 +18,77 @@ public function __construct(Statement $query)
         $this->query = $query;
     }
 
-    public function execute(?array $params = null): bool
+    public function execute($param1, $param2 = null): bool
```

Comment:
why not injecting the connection as `__construct` parameter, for the case that we have an instance

Reply:
That would work as well.

We thought it would be easier when the "deprecation and different parameters" situation is contained in a single method. Otherwise the optional parameter in the constructor would affect a second method and both have to be updated when removing the deprecation 🤷

If you prefer it this way, just says so - I can easily refactor this.

Reply:
the constructor is not affected by backward compatibility. if this is your fear 😉

Reply:
> not affected by backward compatibility

How so? 🤔 It's a public class so I can't just change its contract, can I?
I can make the `Connection` constructor parameter optional which moves the parameter checks and deprecation warning out of the `execute` method into the constructor. If you prefer this version.

Either way the goal should be to deprecate the old "`Connection`-less" usages, so with the next major version `RetryableQuery` will always run with a `Connection` instance to make it deadlock safer.

Reply:
*in _our_ BC strategy 😁

Reply:
I'm sorry if I have to prove you wrong, @mitelg, but that only applies to services in the DI container. 
For classes created at runtime, this is a break.

```
$update = new RetryableQuery(
   $this->connection->prepare('UPDATE promotion SET order_count = :count, orders_per_customer_count = :customerCount WHERE id = :id')
);
```

@hanneswernery I see what you mean. But from the semantic point of view, it would be nicer to pass it in the `__construct`. Can you please change this?

Reply:
sorry, didn't know this is not a service from the container :smile: